shared/lsp/registration.ts

```diff
@@ -363,203 +361,106 @@ export async function register({
-                        throw err
-                    }
-                })()
-                connectionsByRootUri.set(root.uri.toString(), connectionPromise)
-            }
-        }
-        subscriptions.add(
-            from(sourcegraph.workspace.rootChanges)
-                .pipe(
-                    startWith(null),
-                    map(() => [...sourcegraph.workspace.roots]),
-                    scan((before, after) => {
-                        // Create new connections for added workspaces
-                        const added = differenceBy(after, before, root =>
-                            root.uri.toString()
-                        )
-                        addRoots(added)
+                    // Close connections for removed workspaces
+                    const removed = differenceBy(before, after, root =>
+                        root.uri.toString()
+                    )
+                    // tslint:disable-next-line no-floating-promises
+                    Promise.all(
+                        removed.map(async root => {
+                            try {
+                                const connection = await connectionsByRootUri.get(
+                                    root.uri.toString()
+                                )
 
-                        // Close connections for removed workspaces
-                        const removed = differenceBy(before, after, root =>
-                            root.uri.toString()
-                        )
-                        // tslint:disable-next-line no-floating-promises
-                        Promise.all(
-                            removed.map(async root => {
-                                try {
-                                    const connection = await connectionsByRootUri.get(
+                                if (connection) {
+                                    connectionsByRootUri.delete(
                                         root.uri.toString()
                                     )
-                                    if (connection) {
-                                        connection.unsubscribe()
-                                    }
-                                } catch (err) {
-                                    logger.error(
-                                        'Error disposing connection',
-                                        err
-                                    )
+                                    connection.unsubscribe()
                                 }
-                            })
-                        )
-                        return after
-                    })
-                )
-                .subscribe()
-        )
-        addRoots(sourcegraph.workspace.roots)
-        await Promise.all(connectionsByRootUri.values())
-    }
+                            } catch (err) {
+                                logger.error('Error disposing connection', err)
```

Comment:
As 414.

Reply:
This is also unrelated to the PR. We can do this but in a separate effort.